Feat/hierarchical workflow tracing

serglom21 opened the pull request:
## Summary

This PR transforms the GitHub Actions workflow tracing from individual job transactions to a single workflow transaction containing nested job and step spans, providing better visibility into workflow timing and structure.

## Problem that I am attempting to solve

**Before**: Individual job transactions created a flat structure in Sentry, making it difficult to understand workflow timing and relationships.
**After**: Single workflow transaction with proper hierarchical spans showing:
- Workflow-level timing and status
- Individual job spans as children
- Step spans as children of their respective jobs

## Key Changes

**Core Implementation**:
- WorkflowTracer: New class for creating single workflow transactions with nested spans
- WorkflowJobCollector: Collects jobs from workflow runs and sends workflow-level traces
- Enhanced WebAppHandler: Integrates job collection and workflow tracing

## Trace Structure:

```
workflow: Multi-Job Test (transaction)
├─ security-scan (job span)
│  ├─ Set up job (step span)
│  ├─ Checkout code (step span)
│  └─ Run security scan (step span)
├─ performance-tests (job span)
│  ├─ Set up job (step span)
│  ├─ Setup performance test environment (step span)
│  └─ Run performance tests (step span)
└─ backend-tests (job span)
   ├─ Set up job (step span)
   ├─ Setup Python (step span)
   └─ Run backend unit tests (step span)
```

## Files Changed

- `src/workflow_tracer.py` - New WorkflowTracer implementation
- `src/web_app_handler.py` - Enhanced with WorkflowJobCollector
- `src/enhanced_web_app_handler.py` - Alternative handler implementation

## Before

<img width="2705" height="1128" alt="Screenshot 2025-10-01 at 7 13 42 AM" src="https://github.com/user-attachments/assets/57f35d30-7a67-48d9-bf66-41702ade68a2" />

## After

<img width="2723" height="1190" alt="Screenshot 2025-10-01 at 7 14 45 AM" src="https://github.com/user-attachments/assets/a412baed-019b-4658-82d0-7c1f74ba46fa" />

serglom21 committed:
feat: implement hierarchical workflow tracing with single transaction structure

- Add WorkflowTracer class for creating single workflow transactions with nested job/step spans
- Implement WorkflowJobCollector to collect jobs and send workflow-level traces
- Fix Sentry payload structure to match expected format (remove event_id, fix timestamps, correct status mapping)
- Disable individual job transactions to prevent duplicate traces
- Add proper span hierarchy: workflow -> jobs -> steps
- Include trace_version tags for validation
- Add comprehensive logging and error handling
- Disable Flask automatic Sentry tracing to prevent interference
- Add documentation for local development and testing

This change transforms the tracing from individual job transactions to a single
workflow transaction containing all job and step spans, providing better
visibility into workflow timing and structure.

serglom21 committed:
docs: add PR description and update local development guide

serglom21 committed:
refactor: remove markdown docs and revert unnecessary Sentry disabling

- Remove PR_DESCRIPTION.md (will be added manually to PR)
- Revert LOCAL_DEVELOPMENT.md changes
- Restore original Sentry Flask integration in main.py (no interference with WorkflowTracer)
- Restore original github_sdk.py send_trace method (GithubClient not used in current implementation)

serglom21 committed:
docs: remove LOCAL_DEVELOPMENT.md and TESTING_SUMMARY.md

These documentation files are not needed for the core feature implementation.

serglom21 committed:
feat: implement smart workflow completion detection

- Replace hardcoded job count (5) with dynamic thresholds
- Add job arrival time tracking for intelligent processing
- Implement timeout-based detection for small workflows
- Support workflows of any size (1+ jobs) with appropriate timing
- Add cleanup for arrival time tracking data
- Improve logging for better debugging

Smart thresholds:
- 10+ jobs: Process immediately when all complete
- 5-9 jobs: Process when all complete
- 3-4 jobs: Process when all complete
- 1-2 jobs: Process after 3s timeout or immediately if single job

serglom21 committed:
fix: add exception handling to prevent resource leaks in timer callbacks

- Add try-catch block to _process_workflow_immediately method
- Implement _cleanup_workflow_run helper for proper resource cleanup
- Ensure cleanup happens even when exceptions occur in timer threads
- Add comprehensive error logging with stack traces
- Prevent silent failures that could lead to resource leaks

This addresses the Seer bot feedback about unhandled exceptions in timer
callbacks that could cause workflow data to remain in memory indefinitely.

armenzg commented on src/web_app_handler.py:
```diff
@@ -4,21 +4,240 @@
 logger.setLevel(LOGGING_LEVEL)
 
 
+class WorkflowJobCollector:
```
Move this into a new module and import it here.

serglom21 replied:
done!

armenzg commented on src/web_app_handler.py:
```diff
@@ -4,21 +4,240 @@
+                if self._should_process_workflow(run_id, jobs_count):
+                    logger.info(f"Workflow run {run_id} has {jobs_count} jobs, setting timer to process in 2 seconds")
+                    # Set a short timer to allow all jobs to arrive
+                    timer = threading.Timer(2.0, self._process_workflow_immediately, args=[run_id])
```
No magic numbers please: https://refactoring.guru/replace-magic-number-with-symbolic-constant

armenzg commented on src/web_app_handler.py:
```diff
@@ -4,21 +4,240 @@
+            self._cleanup_workflow_run(run_id)
+    
+    def _process_workflow_delayed(self, run_id: int):
+        """Process workflow after delay to allow all jobs to arrive"""
```
What do we do when the jobs do not arrive? or arrive hours later?

serglom21 replied:
good q: I implemented a GitHub API verification to handle jobs that arrive after a workflow is processed. I basically call the API to check the workflow status

armenzg reviewed (changes requested):
Thanks Sergio for the contribution but there are several issues that need addressing before merging this.

The main one is that there are no tests and there are various principles of clean code are not being followed.

You can use Cursor to refactor the PR using "clean code principles".

Also make changes so we can handle the old style and the new style while we verify the changes are working. We may want to only do this for the Sentry org events. We would also need to verify that current CI dashboards in the getsentry repo would not break.

serglom21 committed:
refactor: extract WorkflowJobCollector module with feature flag support

Move WorkflowJobCollector from web_app_handler.py into its own module
with improved documentation, error handling, and feature flag support
for hierarchical tracing. Add comprehensive test coverage.

serglom21 committed:
- Fetch total job count from GitHub API to ensure all jobs are captured
- Use workflow run created_at/updated_at timestamps to match GitHub's duration
- Add cleanup span for workflow teardown time
- Generate GitHub App installation tokens in production
- Increase timeout from 3s to 7s for better job collection

serglom21 committed:
add tests
